internal/receiver/discoveryreceiver/metric_evaluator.go

```diff
@@ -16,26 +16,164 @@ package discoveryreceiver
 
 import (
 	"context"
+	"fmt"
+	"sync"
+	"time"
 
 	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
 	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+
+	"github.com/signalfx/splunk-otel-collector/internal/receiver/discoveryreceiver/statussources"
 )
 
 var _ consumer.Metrics = (*metricEvaluator)(nil)
 
+const (
+	metricMatch = "metric.match"
+)
+
+var (
+	jsonMarshaler = pmetric.NewJSONMarshaler()
+)
+
 // metricEvaluator conforms to a consumer.Metrics to receive any metrics from
 // receiver creator-created receivers and determine if they match any configured
 // Status match rules. If so, they emit log records for the matching metric.
-type metricEvaluator struct{}
+type metricEvaluator struct {
+	*evaluator
+	pLogs chan plog.Logs
+}
 
-func newMetricEvaluator() *metricEvaluator {
-	return &metricEvaluator{}
+func newMetricEvaluator(logger *zap.Logger, cfg *Config, pLogs chan plog.Logs, correlations correlationStore) *metricEvaluator {
+	return &metricEvaluator{
+		pLogs: pLogs,
+		evaluator: &evaluator{
+			logger:        logger,
+			config:        cfg,
+			correlations:  correlations,
+			alreadyLogged: &sync.Map{},
+			// TODO: provide more capable env w/ resource and metric attributes
+			exprEnv: func(pattern string) map[string]any {
+				return map[string]any{"name": pattern}
+			},
+		},
+	}
 }
 
 func (m *metricEvaluator) Capabilities() consumer.Capabilities {
 	return consumer.Capabilities{}
 }
 
-func (m *metricEvaluator) ConsumeMetrics(context.Context, pmetric.Metrics) error {
+func (m *metricEvaluator) ConsumeMetrics(_ context.Context, md pmetric.Metrics) error {
+	if pLogs := m.evaluateMetrics(md); pLogs.LogRecordCount() > 0 {
+		m.pLogs <- pLogs
+	}
 	return nil
 }
+
+func (m *metricEvaluator) evaluateMetrics(md pmetric.Metrics) plog.Logs {
+	if ce := m.logger.Check(zapcore.DebugLevel, "evaluating metrics"); ce != nil {
+		if mbytes, err := jsonMarshaler.MarshalMetrics(md); err == nil {
```

Comment:
Would it be possible to log the error if not nil?

Reply:
updated